README.md
# SilentStorm Integration Helper <!-- omit in toc -->
A [Factorio](https://factorio.com) mod for people streaming the game, that are looking
to spice up their game by letting viewers trigger actions in the game.

The mod does not do anything on its own, without something triggering actions.
To this end the remote interface `silentstorm-integration-helper` is provided by the mod.

You will need to provide your own means of communicating between your streaming platform
of choice and the game. Factorio does provide RCON to interface with the game.

## Table of contents <!-- omit in toc -->
- [Available Functions](#available-functions)
  - [build\_ghosts](#build_ghosts)
    - [Parameters](#parameters)
    - [Examples](#examples)
  - [repair\_base](#repair_base)
    - [Parameters](#parameters-1)
    - [Examples](#examples-1)
  - [build\_bp](#build_bp)
    - [Parameters](#parameters-2)
    - [Examples](#examples-2)
  - [deconstruct](#deconstruct)
    - [Parameters](#parameters-3)
    - [Examples](#examples-3)
  - [gain\_research](#gain_research)
    - [Parameters](#parameters-4)
    - [Examples](#examples-4)
  - [cancel\_research](#cancel_research)
    - [Parameters](#parameters-5)
    - [Examples](#examples-5)
  - [disable\_research](#disable_research)
    - [Parameters](#parameters-6)
    - [Examples](#examples-6)
  - [random\_research](#random_research)
    - [Parameters](#parameters-7)
    - [Examples](#examples-7)
  - [forget\_research](#forget_research)
    - [Parameters](#parameters-8)
    - [Examples](#examples-8)
  - [set\_arti\_speed](#set_arti_speed)
    - [Parameters](#parameters-9)
    - [Examples](#examples-9)
  - [set\_arti\_range](#set_arti_range)
    - [Parameters](#parameters-10)
    - [Examples](#examples-10)
  - [set\_lab\_speed](#set_lab_speed)
    - [Parameters](#parameters-11)
    - [Examples](#examples-11)
  - [teleport](#teleport)
    - [Parameters](#parameters-12)
    - [Examples](#examples-12)
  - [teleport\_distance](#teleport_distance)
    - [Parameters](#parameters-13)
    - [Examples](#examples-13)
  - [enemy\_arty](#enemy_arty)
    - [Parameters](#parameters-14)
    - [Examples](#examples-14)
  - [remove\_entity](#remove_entity)
    - [Parameters](#parameters-15)
    - [Examples](#examples-15)
  - [reset\_recipe](#reset_recipe)
    - [Parameters](#parameters-16)
    - [Examples](#examples-16)
  - [biter\_revive](#biter_revive)
    - [Parameters](#parameters-17)
    - [Examples](#examples-17)
  - [snap\_wires](#snap_wires)
    - [Parameters](#parameters-18)
    - [Examples](#examples-18)
  - [load\_turrets](#load_turrets)
    - [Parameters](#parameters-19)
    - [Examples](#examples-19)
  - [advance\_rocket](#advance_rocket)
    - [Parameters](#parameters-20)
    - [Examples](#examples-20)
  - [rain\_item](#rain_item)
    - [Parameters](#parameters-21)
    - [Examples](#examples-21)
  - [remove\_floor](#remove_floor)
    - [Parameters](#parameters-22)
    - [Examples](#examples-22)
  - [modify\_walk\_speed](#modify_walk_speed)
    - [Parameters](#parameters-23)
    - [Examples](#examples-23)
  - [modify\_craft\_speed](#modify_craft_speed)
    - [Parameters](#parameters-24)
    - [Examples](#examples-24)
  - [modify\_reach\_distance](#modify_reach_distance)
    - [Parameters](#parameters-25)
    - [Examples](#examples-25)
  - [modify\_build\_distance](#modify_build_distance)
    - [Parameters](#parameters-26)
    - [Examples](#examples-26)
  - [modify\_mining\_speed](#modify_mining_speed)
    - [Parameters](#parameters-27)
    - [Examples](#examples-27)
  - [on\_fire](#on_fire)
    - [Parameters](#parameters-28)
    - [Examples](#examples-28)
  - [barrage](#barrage)
    - [Parameters](#parameters-29)
    - [Examples](#examples-29)
  - [dump\_inv](#dump_inv)
    - [Parameters](#parameters-30)
    - [Examples](#examples-30)
  - [cancel\_hand\_craft](#cancel_hand_craft)
    - [Parameters](#parameters-31)
    - [Examples](#examples-31)
  - [start\_hand\_craft](#start_hand_craft)
    - [Parameters](#parameters-32)
    - [Examples](#examples-32)
  - [get\_naked](#get_naked)
    - [Parameters](#parameters-33)
    - [Examples](#examples-33)
  - [vacuum](#vacuum)
    - [Parameters](#parameters-34)
    - [Examples](#examples-34)
  - [drain\_battery](#drain_battery)
    - [Parameters](#parameters-35)
    - [Examples](#examples-35)
  - [drain\_shield](#drain_shield)
    - [Parameters](#parameters-36)
    - [Examples](#examples-36)
  - [drain\_energy](#drain_energy)
    - [Parameters](#parameters-37)
    - [Examples](#examples-37)

## Available Functions

Bold parameters are mandatory and must be provided, otherwise an error will be printed
to the game chat for all players.

All other parameters can be omitted, which will then use the default value.

If you want to specify a later parameter, you need to provide all those before it as well.
You can explicitly opt to use the default value by specifying the parameter as `nil`.

surface for example would usually be something like `game.players['player_name'].surface`
position would similarly be a player position most often e.g. 
`game.players['player_name'].position` but it can also be a fixed position presented like
`{x=0, y=0}`.

### build_ghosts

Builds ghosted entities around a location on the map.
You can optionally have it attempt to build destroyed entities (which have not left
a ghost) as well. However that will not work for all types of buildings due to game
limitations and it may not create anything, or in some cases even create the wrong building.

#### Parameters

- **surface**: A surface of the game.
- **force**: A force of which to repair damaged entities.
- **position**: The center point around which to repair entities.
- range: How many tiles around the position to build entities. Defaults to `20`.
- chance: Percentage chance of a ghosted / destroyed entity actually being built. 
  Defaults to `75`.
- ignore_technology: Boolean. Whether or not to build entities the specified force does
  not yet have researched. Defaults to `false`.
  When set to true it will build stuff that the specified force cannot currently build
  (either because it is not researched, or it cannot be built at all).
- include_remnants: Boolean. Whether or not to attempt to restore destroyed buildings
  that do not have a ghost. Defaults to `false`. When set to true it will _attempt_ 
  to restore destroyed buildings. However due to some limitations no guarantees are made
  as to the success. If a destroyed building cannot be identified it is not build.

#### Examples

Minimum parameters:
`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'build_ghosts', plr.surface, plr.force, plr.position)`

Restore even without ghosts, provided the destroyed building is researched:

`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'build_ghosts', plr.surface, plr.force, plr.position, nil, nil, nil, true)`

### repair_base

The `repair_base` function repairs entities around a location on the map. Minimum and 
maximum health points restored are configurable.

#### Parameters
  - **surface**: A surface of the game on which to repair things.
  - **force**: A force of which to repair damaged entities.
  - **position**: The center point around which to repair entities.
  - range: How many tiles around the position to repair entities. Defaults to `15`.
  - chance: Percentage chance of a damaged entity actually being repaired. Defaults to `75`.
  - minimum_health_gain: If an entity is being repaired it is at least being repaired for
    this many HP. Defaults to `20`.
  - maximum_health_gain: A single entity is never repaired for more than this many HP.
    Defaults to `150`.

#### Examples

Uses the default values for all non-optional parameters. Repairs between 20 and 150 health
for 75% of entities within 15 tiles of the player.
`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'repair_base',
plr.surface, plr.force, plr.position)`

Repairs exactly 100 health for 75% of entities within 10 tiles of the player.
`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'repair_base',
plr.surface, plr.force, plr.position, 10, nil, 100, 100)`

### build_bp

Allows to build arbitrary constructions. Simply create a blueprint of whatever it is
you want to have build, use the share / export feature of the blueprint and copy the
string and specify that as the `blueprint_string` parameter.

By default it will not build a blue belt for example if the force selected didn't yet
research them, but that can be overridden.

#### Parameters
- **surface**: A surface of the game on which to construct the blueprint.
- **force**: A force to use to build the blueprint.
- **position**: The position at which the blueprint build will be centered.
- **blueprint_string**: An exported blueprint string to build.
- ignore_technology: Boolean. Whether or not to actually build entities in the blueprint
  that the chosen force can not currently build. Defaults to `false`. When set to true then
  it will build any entity, even those that have not yet been researched by the force
  specified. When set to false (the default) it will leave such entities as ghosts.

#### Examples

Build the blueprint as yourself centered around player:
`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'build_bp', plr.surface, plr.force, plr.position, '0eNqV0VGLgzAMB/Dv8n/O4Fp13vWrjDF0C1tgRrF1nEi/+7T3cqCD7a0pyS8hmVDfB+560QA3Qc6terjDBC9Xre7LXxg7hoMEbkDQqlkiP9Q+VEFaRSSIXvgXzsQjgTVIEP5TUjCedGhq7ueErXpC13pJz7nbwmQlYYTbmZlWluutbod+AS1lx0gr1r7HFsUL1lC+xWYfTWvKN9n8o2nXbFrCvOh0EPfvfoQH9z4R9tvk5Y8ti735MsbG+ARLmaB2')`

Build a blueprint as enemy:
`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'build_bp', plr.surface, game.forces['enemy'], plr.position, '0eNqNj80Kg0AMhN9lziuY7Y/tvkopRdsgCxplf0pF9t272ovQHnoKEzJfZmY0XeTRWQkwM+x9EA9zmeFtK3W37MI0Mgxs4B4KUveLaqMUITrHAUnByoNfMJSuCizBBssfyiqmm8S+YZcPfvkVxsFnyyDLt4wpdKkw5Uk7Skl9UfR/FNIbSg62FjCbvgpPdn516BPtq7OuDkcqiXRKb62sWyU=')`

### deconstruct

Marks random buildings of the specified force for deconstruction by construction robots,
effectively stopping the building.

#### Parameters
- **surface**: A surface of the game to deconstruct buildings on.
- **force**: A force of which to mark buildings for deconstruction.
- position: A position around which to center the deconstruction. If specified must be
  a valid position around which the range parameter will be applied.
  Defaults to the `entire surface`.
- range: How many tiles around the position to mark buildings for deconstruction. Defaults to `500`.
- chance: Percentage chance of a building being marked for deconstruction. Defaults to `5`.
- maximum_count: How many buildings to mark for deconstruction at most. With a high chance
  and a big base this will cause the impact area to be limited in a small area.
  This function will never mark more than the number of buildings specified here. Defaults to `500`.
- ignore_technology: Boolean. Do not mark for deconstruction anything that the specified force
  cannot currently craft. Defaults to `false`.

#### Examples

`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'deconstruct', plr.surface, plr.force, plr.position, 250)`

### gain_research

Advance the currently active research by an amount

#### Parameters
- **force**: The game force to advance the research for
- chance: Percentage chance to actually advance the research. Defaults to `75`.
- percentage: By what percentage to increase the research by. Valid value range is -100 to 100.
  A value of 100 will finish the research instantly.  
  Negative values will decrease the current research progress. Defaults to a `random value between 10 and 75`.
  Default value has a 5% chance to be negative.

#### Examples

Lose up to 25% of the current research, or gain up to 50% research progress. The actual value
here will vary each time this is ran due to the random number being passed in.

`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'gain_research', plr.force, 100, math.random(-25, 50))`

### cancel_research

Cancels the current active research

#### Parameters
- **force**: The game force to cancel research for
- chance: Percentage chance to actually do something. Defaults to `50`.

#### Examples

`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'cancel_research', plr.force, 100)`

### disable_research
Cancels all the active research for a force, disables all Labs and prevents accessing
the Tech Tree for a specified amount of time.

#### Parameters
- **force**: A game force for which to disable the research.
- chance: Percentage chance for anything to actually happen. Defaults to `50`.
- duration: Number of seconds to disable research for. Defaults to a `random value between 30 and 300 seconds`.

#### Examples

Disable research for 1 minute

`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'disable_research', plr.force, 100, 60)`

### random_research

Start a random research that is available to be researched and put it to the top of the
research queue. If the research queue is full, the last item of the research queue is dropped.

#### Parameters
- **force**: The game force to start a random research for.
- chance: Percentage chance for anything to happen. Defaults to `50`.

#### Examples

`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'random_research', plr.force, 100)`

### forget_research

Removes a random technology that has already been researched. All machines
using the technology lose their recipe. Does **not** disable dependent
technologies. So if you lose Optics for example, you will still be able to
use Laser Turrets, despite it being behind Optics.

#### Parameters
- **force**: The force to disable a random research for.
- chance: Percentage chance to actually do anything. Defaults to `50`.

#### Examples

`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'forget_research', plr.force, 100)`

### set_arti_speed

Increases the artillery shooting speed of a team for a specified number of seconds.

#### Parameters
- **force**: The force to increase the shooting speed for.
- **levels**: How many research levels of artillery speed to add. Valid range: 1 to 21.
- chance: Percentage chance to actually do something. Defaults to `50`
- duration: For how many seconds to increase the shooting speed for. Defaults to a `random value between 60 and 180`.

#### Examples

Increase artillery shooting speed by 5 research levels for 30 seconds

`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'set_arti_speed', plr.force, 5, 100, 30)`

### set_arti_range

Increase the artillery range of a team for a specified number of seconds.

#### Parameters
- **force**: The game force for which to increase the artillery range.
- **levels**: How many research levels of artillery range to add. Valid range: 1 to 21.
- chance: Percentage chance to actually do something. Defaults to `50`.
- duration: For how many seconds to increase the range for. Defaults to a `random value between 60 and 180`.

#### Examples

Increase artillery shooting range by 5 research levels for 30 seconds

`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'set_arti_range', plr.force, 5, 100, 30)`

### set_lab_speed

Increases the speed that research labs work at. This will make research progress
faster, however it will also make them consume science packs faster. Can be considered
the less cheaty version of [gain_research](#gainresearch).

#### Parameters
- **force**: The game force for which to increase the lab speed.
- change: percentage by which to change the lab speed. -100 will make labs run half their
  current speed. 100 will make labs run twice their current speed.  
  Defaults to `a random value between 1 and 100`. Default value has a 5% chance to be negative.
  Valid range is -100 to 100.
- chance: Percentage chance for lab speed change to actually happen. Defaults to `75`.
- duration:  (random 10 - 180)

#### Examples

Run all the labs of the specified force, twice as fast for 60 seconds:

`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'set_lab_speed', plr.force, 100, 100, 60)`

### teleport

Teleports a player to the given location on a given surface after a given delay.
If the player is inside a vehicle it will be teleported alongside the player if
possible, otherwise the player is evicted from the vehicle before teleport.

Will not teleport player into places they cannot get out of.
Will not teleport player onto water if they cannot swim.

#### Parameters
- **player**: The player to teleport
- **position**: The position to teleport the player to.
- delay: Seconds before the teleport happens. Defaults to a `random value between 1 and 10`.
- surface: The destination surface to teleport the player to. Defaults to `their current surface`.

#### Examples

Teleport the player to coordinates 250, -500 onto Nauvis Orbit after 10 seconds.

`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'teleport', plr, {x=250, y=-500}, 10, game.surfaces['Nauvis Orbit'])`

### teleport_distance

Teleports a player the given number of tiles from their current location.
If the player is inside a vehicle it will be teleported alongside the player if
possible, otherwise the player is evicted from the vehicle before teleport.

Will not teleport player into places they cannot get out of.
Will not teleport player onto water if they cannot swim.

If a surface is specified, please note that the player coordinates on their **current** surface
will be used as _origin_ coordinates. If the specified surface does not contain these coordinates
the teleport will fail.

#### Parameters
- **player**: The player to teleport
- **distance**: Number of tiles to teleport the player.
- delay: Number of seconds to delay the teleport for. Defaults to a `random value from 1 to 10` seconds.
- surface: The surface to teleport the player to. Defaults to `their current surface`.

#### Examples

After 5 seconds teleport the player 250 tiles from their current location.

`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'teleport_distance', plr, 250, 5)`

### enemy_arty

Finds artillery turrets of the provided force and converts them to an enemy artillery turret.

#### Parameters
- **surface**: The surface to search for artillery turrets.
- **force**: The game force (team) of which to find artillery turrets.
- position: The position to center the search area around. Defaults to `{x=0, y=0}`.
- range: Number of tiles around the position to find artillery turrets.
  Defaults to a `random value between 500 and 5000`.
- max: Maximum number of artillery turrets to convert to an enemy artillery.
  Defaults to a `random value between 1 and 10`.
- chance: Percentage chance to convert a found artillery turret to an enemy turret.
  Defaults to a `random value between 10 and 100`.
- enemy_force: The force to assign the artillery turret to. Defaults to `game.forces['enemy']`.

#### Examples

Find up to 2 artillery turrets of the specified players force within 500 tiles around their
position and convert them to the enemy force.

`local plr = game.players['foo'] remote.call('silentstorm-integration-helper', 'enemy_arty', plr.surface, plr.force, plr.position, 500, 2, 100)`

### remove_entity

Outright erases buildings from the map. Can either erase random or specific entities.
When removing random entities it will pick a random building to remove and remove up
to the specified number of this entity. It will **NOT** remove multiple different entities.

#### Parameters
- **surface**: The surface to search for buildings to remove
- **force**: The force of which to delete buildings
- position: The center of the area to search for buildings to remove. Defaults to `{x=0, y=0}`.
- range: The number of tiles around the position to search for buildings to remove.
  Defaults to a `random value between 40 and 200`.

changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.4.16
Date: 2023-08-07

  Bugfixes:
    - remove_floor overwrite hidden tile for removed floor tiles.
---------------------------------------------------------------------------------------------------
Version: 0.4.15
Date: 2023-08-06

  Changes:
    - remove_floor now checks for valid tile choices if specified and will error out when an
      invalid name is specified.
---------------------------------------------------------------------------------------------------
Version: 0.4.14
Date: 2023-08-06

  Features:
    - Added remove_floor remote call. Removes player placeable tiles (Concrete, Stone path, etc)

---------------------------------------------------------------------------------------------------
Version: 0.4.13
Date: 2023-07-10

  Bugfixes:
    - Fixed rain_item sometimes reporting decimals due to calculation errors.

---------------------------------------------------------------------------------------------------
Version: 0.4.12
Date: 2023-05-27

  Bugfixes:
    - Fixed homing count default value calculation for barrage command.

---------------------------------------------------------------------------------------------------
Version: 0.4.11
Date: 2022-12-24

  Bugfixes:
    - Fixed unrecoverable error when player died in between vacuum starting and ending but was
      already alive when it ended.

---------------------------------------------------------------------------------------------------
Version: 0.4.10
Date: 2022-12-18

  Features:
    - Added ability to modify mining speed of a player

---------------------------------------------------------------------------------------------------
Version: 0.4.9
Date: 2022-12-08

  Bugfixes:
    - Fixed unrecoverable error when player is dead when armor removal times out and attempts
      to return the armor to the player.
    - Added a few more safety checks related to player death.

---------------------------------------------------------------------------------------------------
Version: 0.4.8
Date: 2022-11-29

  Features:
    - Added item_class and all_players parameters to remove_entity command. Allows centering
      entity removal around all players. item_class overrides the item name if specified and
      can be used to find any entity of a specific type rather than a specific entity type.
      Use item_class 'assembling-machine' rather than 'assembling-machine-1' as name to remove
      any assembling machine, rather than just Tier 1 Assemblers.

---------------------------------------------------------------------------------------------------
Version: 0.4.7
Date: 2022-11-28

  Bugfixes:
    - Fix teleport being delayed until a location could be found when the spawn location
      was modified to not be at 0,0 and there was no valid location around 0,0.

  Features:
    - Allow remove_entity to leave the contents of removed entities rather than removing it.

---------------------------------------------------------------------------------------------------
Version: 0.4.6
Date: 2022-11-11

  Features:
    - Allow to increase the build distance of a player.
    - Allow to increase the reach distance of a player.

---------------------------------------------------------------------------------------------------
Version: 0.4.5
Date: 2022-11-04

  Bugfixes:
    - Fixed unrecoverable error when player dies while affected by vacuum command.
    - Fixed ramp_duration not working for drain_* commands without also specifying
      a non-zero delay.

---------------------------------------------------------------------------------------------------
Version: 0.4.4
Date: 2022-11-01

  Bugfixes:
    - Fixed drain_battery and drain_shield not respecting a 0 value for absolutes

  Features:
    - Added drain_energy command that drains all equipment energy
    - Added a ramp_duration parameter to drain_* commands, to adjust the value
      over time, rather than instantly setting the desired value.

---------------------------------------------------------------------------------------------------
Version: 0.4.3
Date: 2022-10-15

  Bugfixes:
    - Fixed accessing an undefined variable in modify_walk_speed and modify_craft_speed
    - Fixed more unknown key errors in text messages
    - Fixed an error when calling load_turrets

---------------------------------------------------------------------------------------------------
Version: 0.4.2
Date: 2022-10-07

  Bugfixes:
    - Fixed an unintended side effect of 0.4.1 changes to barrage, which caused
      it to no longer fire when the player targeted died.

  Changes:
    - Barrage command can now receive an optional position and surface to target
      rather than targeting the player.
    - Barrage command now has an optional parameter that allows to specify whether
      or not to continue shooting when the player targeted is dead. For consistency
      with previous versions this is enabled by default.

---------------------------------------------------------------------------------------------------
Version: 0.4.1
Date: 2022-10-07

  Bugfixes:
    - Fixed an issue which caused names to not be displayed correctly
    - Fixed an issue with vacuum command causing an unrecoverable error when player dies
      during its lifetime.
    - Fixed a typo in the default message for cancel_handcraft

  Features:
    - Improved compatibility with Space Exploration Navigation Satellite mode. Calls will
      now target the actual player, instead of the current navigation satellite view.

---------------------------------------------------------------------------------------------------
Version: 0.4.0
Date: 2022-09-09

  Important:
    - You will need to change your calls as some parameters have changed.
    - Please read the entire changelog of this version carefully!

  Changes:
    - the order of parameters for the teleportation commands has changed. The surface the

funcs/build.lua
-- Copyright 2022 Sil3ntStorm https://github.com/Sil3ntStorm
--
-- Licensed under MS-RL, see https://opensource.org/licenses/MS-RL

local build = {}
local research = require('funcs/research_tree')
local constants = require('constants')
local config = require('utils/config')
local strutil = require('utils/string_replace')
local proto = require('utils/proto')
local fml = require('utils/lua_is_stupid')
local map = require('funcs/map')
local tc = require('utils/type_check')

local function createResearchedGhosts(entities, chance, ignore_tech)
    local count = 0
    for _, e in pairs(entities) do
        if ignore_tech or research.can_build(e.force, e.ghost_name) then
            if math.random(1,100) <= chance then
                e.revive()
                count = count + 1
            end
        end
    end
    return count
end

local function getArea(entities)
    local x = nil
    local y = nil
    local x2 = nil
    local y2 = nil
    local padding = 0.6
    for _, e in pairs(entities) do
        if not x then
            x = math.ceil(e.position.x - padding)
        end
        if not x2 then
            x2 = math.floor(e.position.x + padding)
        end
        if not y then
            y = math.ceil(e.position.y - padding)
        end
        if not y2 then
            y2 = math.floor(e.position.y + padding)
        end
        if x > e.position.x then
            x = math.ceil(e.position.x - padding)
        end
        if x2 < e.position.x then
            x2 = math.floor(e.position.x + padding)
        end
        if y > e.position.y then
            y = math.ceil(e.position.y - padding)
        end
        if y2 < e.position.y then
            y2 = math.floor(e.position.y + padding)
        end
    end
    return {left_top={x, y}, right_bottom={x2, y2}}
end

build.build_ghosts = function(surface, force, position, range, chance, ignore_tech, include_remnants)
    if not tc.is_surface(surface) or not tc.is_force(force) or not tc.is_position(position) then
        game.print('Invalid input parameters. Surface, Force, Position are required', constants.error)
        return
    end
    if ignore_tech ~= false and ignore_tech ~= true then
        ignore_tech = false
    end
    if include_remnants ~= false and include_remnants ~= true then
        include_remnants = false
    end
    range = range or 20
    chance = chance or 75
    local count = createResearchedGhosts(surface.find_entities_filtered{name='entity-ghost', radius=range, position=position, force=force}, chance, ignore_tech)
    local validTypes = proto.get_entity_prototypes()
    if include_remnants then
        local corpses = surface.find_entities_filtered{type='corpse', radius=range, position=position}
        for _, corpse in pairs(corpses) do
            if corpse.valid then
                local toBuild = string.sub(corpse.name, 1, -10)
                local isValidType = fml.contains(validTypes, toBuild)
                if not isValidType then
                    log('Type ' .. toBuild .. ' is not listed in valid entities: ' .. serpent.line(validTypes))
                end
                if math.random(1, 100) <= chance and string.sub(corpse.name, -8) == 'remnants' and (ignore_tech or research.can_build(force, toBuild)) and isValidType then
                    local res = surface.create_entity{
                        name = toBuild,
                        position = corpse.position,
                        direction = corpse.direction,
                        raise_built = true,
                        move_stuck_players = true,
                        force = force
                    }
                    if res then
                        count = count + 1
                    end
                end
            end
        end
        -- Handle rails, which for some stupid reason are special
        local entities = surface.find_entities_filtered{type='rail-remnants', radius=range, position=position}
        for _, ent in pairs(entities) do
            if ent.valid then
                local toBuild = string.sub(ent.name, 1, -10)
                local isValidType = fml.contains(validTypes, toBuild)
                if math.random(1, 100) <= chance and string.sub(ent.name, -8) == 'remnants' and (ignore_tech or research.can_build(force, toBuild)) and isValidType then
                    local res = surface.create_entity{
                        name = toBuild,
                        position = ent.position,
                        direction = ent.direction,
                        raise_built = true,
                        move_stuck_players = true,
                        force = force
                    }
                    if res then
                        count = count + 1
                    end
                end
            end
        end
    end

    if #config['msg-build-ghost'] > 0 then
        force.print(strutil.replace_variables(config['msg-build-ghost'], {count}), count > 0 and constants.good or constants.neutral)
    end

end

build.repair_base = function(surface, force, position, range, chance, min_gain, max_gain)
    if not tc.is_surface(surface) or not tc.is_force(force) or not tc.is_position(position) then
        game.print('Invalid input parameters. Surface, Force, Position are required', constants.error)
        return
    end
    range = range or 15
    chance = chance or 75
    min_gain = min_gain or 20
    max_gain = max_gain or 150
    local total_gain = 0
    local count = 0
    for _, e in pairs(surface.find_entities_filtered{radius=range, position=position, force=force}) do
        if e.name ~= 'entity-ghost' and e.type ~= 'fire' and e.type ~= 'character' and not e.health then
            log('Attempted to repair entity without health:' .. e.type .. ' ' .. e.name)
        end
        if e.name ~= 'entity-ghost' and e.type ~= 'fire' and e.type ~= 'character' and e.health ~= nil and math.random(1,100) <= chance then
            local max = math.ceil(e.health / e.get_health_ratio())
            local heal = math.random(math.max(0, min_gain), math.max(min_gain, max_gain))
            local new_health = math.min(e.health + heal, max)
            if new_health - e.health > 0 then
                count = count + 1
            end
            total_gain = total_gain + (new_health - e.health)
            e.health = new_health
        end
    end
    if #config['msg-build-repair'] > 0 then
        force.print(strutil.replace_variables(config['msg-build-repair'], {count, math.floor(total_gain + 0.5)}), constants.good)
    end
end

build.build_blueprint = function(surface, force, position, bp_string, ignore_tech)
    if not tc.is_surface(surface) or not tc.is_force(force) or not tc.is_position(position) or not bp_string or #bp_string < 10 then
        game.print('Invalid input parameters. Surface, Force, Position, Blueprint are required', constants.error)
        return
    end
    if ignore_tech ~= false and ignore_tech ~= true then
        ignore_tech = false
    end
    local bp = surface.create_entity{name='item-on-ground', position=map.getRandomPositionInRange(position, 90), stack='blueprint'}
    if bp.stack.import_stack(bp_string) == 1 then
        game.print('Invalid Blueprint', constants.error)
        bp.destroy()
        return
    end
    local entities = bp.stack.build_blueprint{surface=surface, force=force, position=position, force_build=true, skip_fog_of_war=false, raise_built=false}
    bp.destroy()
    if #entities > 0 then
        local area = getArea(entities)
        local destroy_cliffs = ignore_tech or research.can_build(force, 'cliff-explosives')
        for _, e in pairs(surface.find_entities_filtered{area=area, to_be_deconstructed=true}) do
            if e.type == 'tree' or string.sub(e.name, 1, 5) == 'rock-' then
                e.destroy()
            elseif e.type == 'cliff' and destroy_cliffs then
                e.destroy()
            end
        end
    end
    local count = createResearchedGhosts(entities, 100, ignore_tech)
    if #config['msg-build-bp'] > 0 then
        force.print(strutil.replace_variables(config['msg-build-bp'], {count}), count > 0 and constants.good or constants.neutral)
    end
end

function build.deconstruct(surface, force, position, range, chance, max_count, ignore_tech)
    if not tc.is_surface(surface) or not tc.is_force(force) then
        game.print('Invalid input parameters. Surface and Force are required', constants.error)
        return
    end
    if position and not tc.is_position(position) then
        game.print('Invalid input parameters. If specified position must be valid position', constants.error)
        return
    end
    range = range or 500
    chance = chance or 5
    max_count = max_count or 2048
    if ignore_tech ~= false and ignore_tech ~= true then
        ignore_tech = false
    end

    chance = math.max(0, math.min(chance, 100))
    chance = chance * 10

    local entities = {}
    if tc.is_position(position) and type(range) == 'number' then
        entities = surface.find_entities_filtered{position=position, radius=range, force=force, to_be_deconstructed=false}
    else
        entities = surface.find_entities_filtered{force=force, to_be_deconstructed=false}
    end

    local count = 0
    for _, ent in pairs(entities) do
        if count < max_count and math.random(1, 1000) <= chance and (ignore_tech or research.can_build(force, ent.name)) then
            if ent.order_deconstruction(force) then
                count = count + 1
            end
        end
    end

    if #config['msg-build-decon'] > 0 then
        force.print(strutil.replace_variables(config['msg-build-decon'], {count}), count > 0 and constants.bad or constants.neutral)
    end
end

return build
